drivers/storage/nvme/NvmeProbe.c

```diff
@@ -0,0 +1,645 @@
+*/
+BOOLEAN
+NVMeInitNS(PNVME_DEVICE_EXTENSION pDevExt)
+{
+    if (pDevExt == NULL)
+        return FALSE;
+
+    BOOLEAN foundNS = FALSE;
+    NVMe_COMMAND nvmeCmd = {0};
+    NVMe_COMPLETION_QUEUE_ENTRY cmpEntry = {0};
+
+    /* Lets see if we have any attached NS */
+    nvmeCmd.CDW0.OPC = ADMIN_IDENTIFY;
+    nvmeCmd.PRP1     = pDevExt->QueueInfo.SubQueueInfo->PRPListStartPhysical.QuadPart;
+    ((PADMIN_IDENTIFY_COMMAND_DW10)&nvmeCmd.CDW10)->CNS = CNS_LIST_ATTACHED_NAMESPACES;
+
+    RtlZeroMemory(pDevExt->QueueInfo.SubQueueInfo->PRPListStartVirtual, PAGE_SIZE);
+
+    if (!NVMeIssueCmd(pDevExt, &nvmeCmd, &cmpEntry, FALSE)) {
+        DbgPrint("Failed to issue List attach NS\n");
+        return FALSE;
+    }
+
+    if (!NVMeIsCmdSuccessful(&cmpEntry)) {
+        DbgPrint("Command Identify failed CNS:0x%x\n", CNS_LIST_ATTACHED_NAMESPACES);
+        return FALSE;
+    }
+
+    PUINT32 pNSList = (PUINT32)pDevExt->QueueInfo.SubQueueInfo->PRPListStartVirtual;
+    for (int i = 0; i < MAX_NAMESPACES; i++){
+        if (pNSList[i] == 0)
+            break;
+        
+        /* Valid NS, lets store it */
+        pDevExt->NSInfo.namespaceId = pNSList[i];
+        pDevExt->NSInfo.nsStatus = NS_ATTACHED;
+        foundNS = TRUE;
+    }
+
+    if (foundNS == FALSE) {
+        /* TODO: Check if we have allocated NS but not attached; else create NS */
+        DbgPrint("No Active NS found for controller %s\n", pDevExt->controllerIdentifyData.SN);
+        return FALSE;
+    }
+
+    /* Reset */
+    RtlZeroMemory(&nvmeCmd, sizeof(NVMe_COMMAND));
+    RtlZeroMemory(pDevExt->QueueInfo.SubQueueInfo->PRPListStartVirtual, sizeof(NVMe_COMMAND));
+
+    /* We have a NS created and attached. Lets id the NS */
+    nvmeCmd.CDW0.OPC = ADMIN_IDENTIFY;
+    nvmeCmd.NSID     = pDevExt->NSInfo.namespaceId;
+    nvmeCmd.PRP1     = pDevExt->QueueInfo.SubQueueInfo->PRPListStartPhysical.QuadPart;
+    ((PADMIN_IDENTIFY_COMMAND_DW10)&nvmeCmd.CDW10)->CNS = CNS_IDENTIFY_NAMESPACE;
+
+    if (!NVMeIssueCmd(pDevExt, &nvmeCmd, &cmpEntry, FALSE)) {
+        DbgPrint("Failed to issue Identify Namespace for NSID:0x%x\n", pDevExt->NSInfo.namespaceId);
+        return FALSE;
+    }
+
+    if (!NVMeIsCmdSuccessful(&cmpEntry)) {
+        DbgPrint("Command Identify failed CNS:0x%x\n", CNS_IDENTIFY_NAMESPACE);
+        return FALSE;
+    }
+
+    RtlCopyMemory(&pDevExt->NSInfo.identifyData, pDevExt->QueueInfo.SubQueueInfo->PRPListStartVirtual, sizeof(ADMIN_IDENTIFY_NAMESPACE));
+    DbgPrint("Found NS[0x%x] NSZE 0x%llx NCAP 0x%llx\n", pDevExt->NSInfo.namespaceId, pDevExt->NSInfo.identifyData.NSZE, pDevExt->NSInfo.identifyData.NCAP);
+
+    pDevExt->NSInfo.BlockSize = 1 << pDevExt->NSInfo.identifyData.LBAFx[pDevExt->NSInfo.identifyData.FLBAS.SupportedCombination].LBADS;
+    pDevExt->NSInfo.IsNamespaceReadOnly = FALSE;
+    pDevExt->NSInfo.nsReady = TRUE;
+
+    return TRUE;
+}
+
+BOOLEAN
+NVMeInitIOQueues(PNVME_DEVICE_EXTENSION pDevExt)
+{
+    if (pDevExt == NULL)
```

Comment:
Why/how would this be null? Same question for the two other functions above.

Reply:
Its not currently but this is not guaranteed to be only called from the same function. Also checking passed parameters is always a good thing when it comes to writting C.
Its a personal prefrence really. And dosen't hurt